examples/ex9.cpp

```diff
@@ -51,6 +55,1094 @@ double inflow_function(const Vector &x);
+      }
+   }
+
+private:
+   
+   double distance_(const IntegrationPoint &a, const IntegrationPoint &b)
+   {
+      return sqrt((a.x - b.x) * (a.x - b.x) +
+                  (a.y - b.y) * (a.y - b.y) +
+                  (a.z - b.z) * (a.z - b.z));
+   }
+   
+   double Distance(const int dof1, const int dof2) const
+   {
+      const int dim = fes->GetMesh()->Dimension();
+      
+      if (dim==1)
+      {
+         return abs(DOFs_coord(0, dof1) - DOFs_coord(0, dof2));
+      }
+      else if (dim==2)
+      {
+         const double d1 = DOFs_coord(0, dof1) - DOFs_coord(0, dof2),
+         d2 = DOFs_coord(1, dof1) - DOFs_coord(1, dof2);
+         return sqrt(d1*d1 + d2*d2);
+      }
+      else
+      {
+         const double d1 = DOFs_coord(0, dof1) - DOFs_coord(0, dof2),
+         d2 = DOFs_coord(1, dof1) - DOFs_coord(1, dof2),
+         d3 = DOFs_coord(2, dof1) - DOFs_coord(2, dof2);
+         return sqrt(d1*d1 + d2*d2 + d3*d3);
+      }
+   }
+   
+   // Fills DOFs_coord
+   void ComputeCoordinates(FiniteElementSpace *fes)
+   {
+      const int dim = fes->GetMesh()->Dimension();
+      const int num_cells = fes->GetNE();
+      const int NDOFs     = fes->GetVSize();
+      DOFs_coord.SetSize(dim, NDOFs);
+      // DOFs_coord.SetSize(dim, NDOFs + fes->num_face_nbr_dofs); // for parallel
+      
+      Array<int> ldofs;
+      DenseMatrix physical_coord;
+      
+      // Cells for the current process.
+      for (int i = 0; i < num_cells; i++)
+      {
+         const IntegrationRule &ir = fes->GetFE(i)->GetNodes();
+         ElementTransformation *el_trans = fes->GetElementTransformation(i);
+         
+         el_trans->Transform(ir, physical_coord);
+         fes->GetElementDofs(i, ldofs);
+         
+         for (int j = 0; j < ldofs.Size(); j++)
+         {
+            for (int d = 0; d < dim; d++)
+            {
+               DOFs_coord(d, ldofs[j]) = physical_coord(d, j);
+            }
+         }
+      }
+      
+      // Face-neighbor cells.
+      /* for parallel
+      IsoparametricTransformation el_trans;
+      for (int i = 0; i < fes->GetMesh()->face_nbr_elements.Size(); i++)
+      {
+         const IntegrationRule &ir = fes->GetFaceNbrFE(i)->GetNodes();
+         fes->GetMesh()->GetFaceNbrElementTransformation(i, &el_trans);
+         
+         el_trans.Transform(ir, physical_coord);
+         fes->GetFaceNbrElementVDofs(i, ldofs);
+         
+         for (int j = 0; j < ldofs.Size(); j++)
+         {
+            ldofs[j] += NDOFs;
+            
+            for (int d = 0; d < dim; ++d)
+            {
+               DOFs_coord(d, ldofs[j]) = physical_coord(d, j);
+            }
+         }
+      } */
+   }
+   
+   void GetAllNeighbors(const SparseMatrix& K)
+   {
+      const int *I = K.GetI(), *J = K.GetJ(), size = K.Size();
+      
+      for (int i = 0, k = 0; i < size; i++)
+      {
+         for (int end = I[i+1]; k < end; k++)
+         {
+            map_for_SmoothnessIndicator[i].push_back(J[k]);
```

Comment:
Is this map going to be used for anything?

Reply:
Not at the moment.

Reply:
Removed it.